servers/src/main/java/tachyon/worker/block/TieredBlockStore.java

```diff
@@ -0,0 +1,432 @@
+/*
+ * Licensed to the University of California, Berkeley under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.worker.block;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+import tachyon.Constants;
+import tachyon.Pair;
+import tachyon.conf.TachyonConf;
+import tachyon.thrift.InvalidPathException;
+import tachyon.util.CommonUtils;
+import tachyon.worker.block.allocator.Allocator;
+import tachyon.worker.block.allocator.AllocatorFactory;
+import tachyon.worker.block.allocator.AllocatorType;
+import tachyon.worker.block.evictor.EvictionPlan;
+import tachyon.worker.block.evictor.Evictor;
+import tachyon.worker.block.evictor.EvictorType;
+import tachyon.worker.block.evictor.EvictorFactory;
+import tachyon.worker.block.io.BlockReader;
+import tachyon.worker.block.io.BlockWriter;
+import tachyon.worker.block.io.LocalFileBlockReader;
+import tachyon.worker.block.io.LocalFileBlockWriter;
+import tachyon.worker.block.meta.BlockMeta;
+import tachyon.worker.block.meta.TempBlockMeta;
+
+/**
+ * This class represents an object store that manages all the blocks in the local tiered storage.
+ * This store exposes simple public APIs to operate blocks. Inside this store, it creates an
+ * Allocator to decide where to put a new block, an Evictor to decide where to evict a stale block,
+ * a BlockMetadataManager to maintain the status of the tiered storage, and a LockManager to
+ * coordinate read/write on the same block.
+ * <p>
+ * This class is thread-safe.
+ */
+public class TieredBlockStore implements BlockStore {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private final TachyonConf mTachyonConf;
+  private final BlockMetadataManager mMetaManager;
+  private final BlockLockManager mLockManager;
+  private final Allocator mAllocator;
+  private final Evictor mEvictor;
+  private List<BlockStoreEventListener> mBlockStoreEventListeners =
+      new ArrayList<BlockStoreEventListener>();
+
+  /** A readwrite lock for meta data **/
```

Comment:
readwrite -> read/write

Also, this means the lock is protecting the metadata WITH the data? Is this for atomicity?

Do all read-only access to the metadata need the read lock? Because some metadata reads are not being protected by this lock. Then, in which situations must the read lock be held, and which situations must the write lock be held?